### ISSUE TITLE
[Mono] Input GetState() causes abnormal program termination

### ISSUE THREAD

**Issue by derrickcreamer** [opentk/opentk#274]
(Tested on Mono + Windows 7. I can test on Mono + Linux soon.)

Any call to OpenTK.Input.Keyboard.GetState() or OpenTK.Input.Mouse.GetState() (on Mono + Windows) causes unwanted effects when the program ends:

Behavior:
If there is no window, the program simply keeps running in the background until killed.
Expected:
The program should end & close once it reaches the end of Main or Environment.Exit(0) is called.

Behavior:
If a window is open, window.Close() or window.Exit() can fail after GetState() has been called. In any case, after the program ends, if a window is still open, Mono will almost always crash after a few seconds.
Expected:
Close() and Exit() should succeed. Program should end cleanly.

I'll emphasize that GetState() itself works as expected when called, and KeyboardState methods such as IsKeyDown also work correctly. The problem only surfaces when the program tries to end.

Code to reproduce:

``` c#
using OpenTK;
namespace GetStateBug{
 public static class GetStateMain{
  public static void Main(){
   OpenTK.Input.Keyboard.GetState(); // After execution, note that mono.exe is still running.
  } // (and your IDE will offer the 'Stop' action)
 }
}
```

Code to recreate this bug with an open window:

``` c#
using System;
using OpenTK;
using OpenTK.Graphics;
using OpenTK.Graphics.OpenGL;
namespace TKGetStateBugReport{
    public static class BugReportMain{
        public static void Main(){
            GameWindow w = new GameWindow(200,100,GraphicsMode.Default,"Testing");
            w.RenderFrame += (sender, e) => {
                GL.ClearColor(0.4f,0.0f,0.0f,1.0f);
                GL.Clear(ClearBufferMask.ColorBufferBit);
                w.SwapBuffers();
            };
            w.UpdateFrame += (sender, e) => {
// Pressing A here will cause the window to remain open and unresponsive,
// crashing Mono a few seconds after you try to interact with the window
                if(OpenTK.Input.Keyboard.GetState().IsKeyDown(OpenTK.Input.Key.A)){
                    w.Exit();
                    Environment.Exit(0);
                }
            };
            w.Run();
        }
    }
}
```

**Comment by glopesdev** [comment]
This is probably related to the background thread that is created to update the input state. This thread is started on the first time you try to retrieve any input state and keeps going forever.

Did you try calling Toolkit.Dispose()? It should in principle destroy all shared resources including the input drivers, although I'm also having trouble shutting down everything cleanly. In fact, even looking at the source code it's hard to follow the call graph of methods that lead to disposal of the input driver.

@Frassle @thefiddler Is there a way to properly shutdown the input driver thread? I'm also having issues in completely shutting down OpenTK, especially if you want to bring it back later. It looks like most of OpenTK's platform initialization system is built on global static variables which get initialized once and cached forever, which effectively makes it impossible to "kill" without bringing down the whole process.

**Comment by Frassle** [comment]
There's a lot of stuff that probably doesn't handle shutdown/startup cycles well. I'll try to have a look through and of course any cleanups anyone else finds pullreqs welcome.

**Comment by varon** [comment]
Closing as this is an ancient issue relating to a very old version of mono on windows, which isn't a supported configuration.
